File: integrations/langchain-js/minigraf-chat-history.mjs
import { BaseChatMessageHistory } from "@langchain/core/chat_history";
import { AIMessage, HumanMessage } from "@langchain/core/messages";
import minigraf from "minigraf";

const { MiniGrafDb } = minigraf;

export const EXPECTED_OUTPUT = `Human: Remember that Minigraf stores agent memory.
AI: Got it. I will use Minigraf-backed chat history.`;

function datalogString(value) {
  return JSON.stringify(value);
}

function sessionSlug(value) {
  return value.replace(/[^A-Za-z0-9_-]+/g, "-").replace(/^-+|-+$/g, "") || "default";
}

function messageRole(message) {
  if (message instanceof HumanMessage) {
    return "human";
  }
  if (message instanceof AIMessage) {
    return "ai";
  }
  return message.getType();
}

function roleToMessage(role, content) {
  if (role === "human") {
    return new HumanMessage(content);
  }
  if (role === "ai") {
    return new AIMessage(content);
  }
  throw new Error(`unsupported message role: ${role}`);
}

export class MinigrafChatMessageHistory extends BaseChatMessageHistory {
  constructor({ sessionId, db = MiniGrafDb.inMemory() }) {
    super();
    this.sessionId = sessionId;
    this.sessionSlug = sessionSlug(sessionId);
    this.db = db;
  }

  async getMessages() {
    const session = datalogString(this.sessionId);
    const result = JSON.parse(
      this.db.execute(`(query [:find ?m ?index ?role ?content
                               :where [?m :message/session ${session}]
                                      [?m :message/index ?index]
                                      [?m :message/role ?role]
                                      [?m :message/content ?content]])`),
    );
    return result.results
      .sort((left, right) => left[1] - right[1])
      .map(([, , role, content]) => roleToMessage(role, content));
  }

  async addMessages(messages) {
    const start = (await this.getMessages()).length;
    const facts = messages.flatMap((message, offset) => {
      const index = start + offset;
      const entity = `:message/${this.sessionSlug}-${index}`;
      return [
        `[${entity} :message/session ${datalogString(this.sessionId)}]`,
        `[${entity} :message/index ${index}]`,
        `[${entity} :message/role ${datalogString(messageRole(message))}]`,
        `[${entity} :message/content ${datalogString(String(message.content))}]`,
      ];
    });
    if (facts.length > 0) {
      this.db.execute(`(transact [${facts.join(" ")}])`);
    }
  }

  async addMessage(message) {
    await this.addMessages([message]);
  }

  async clear() {
    const session = datalogString(this.sessionId);
    const result = JSON.parse(
      this.db.execute(`(query [:find ?m ?index ?role ?content
                               :where [?m :message/session ${session}]
                                      [?m :message/index ?index]
                                      [?m :message/role ?role]
                                      [?m :message/content ?content]])`),
    );
    const facts = result.results.flatMap(([messageId, index, role, content]) => {
      const entity = `#uuid "${messageId}"`;
      return [
        `[${entity} :message/session ${session}]`,
        `[${entity} :message/index ${index}]`,
        `[${entity} :message/role ${datalogString(role)}]`,
        `[${entity} :message/content ${datalogString(content)}]`,
      ];
    });
    if (facts.length > 0) {
      this.db.execute(`(retract [${facts.join(" ")}])`);
    }
  }
}

export async function demo() {
  const history = new MinigrafChatMessageHistory({ sessionId: "agent-demo" });
  await history.addMessages([
    new HumanMessage("Remember that Minigraf stores agent memory."),
    new AIMessage("Got it. I will use Minigraf-backed chat history."),
  ]);

  for (const message of await history.getMessages()) {
    const label = message instanceof HumanMessage ? "Human" : "AI";
    console.log(`${label}: ${message.content}`);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  await demo();
}

File: integrations/langchain-python/minigraf_chat_history.py
from __future__ import annotations

import json
import re
import tempfile
from pathlib import Path
from typing import Sequence

from langchain_core.chat_history import BaseChatMessageHistory
from langchain_core.messages import AIMessage, BaseMessage, HumanMessage
from minigraf import MiniGrafDb


EXPECTED_OUTPUT = """Human: Remember that Minigraf stores agent memory.
AI: Got it. I will use Minigraf-backed chat history."""


def _datalog_string(value: str) -> str:
    return json.dumps(value)


def _session_slug(value: str) -> str:
    slug = re.sub(r"[^A-Za-z0-9_-]+", "-", value).strip("-")
    return slug or "default"


def _message_to_role(message: BaseMessage) -> str:
    if isinstance(message, HumanMessage):
        return "human"
    if isinstance(message, AIMessage):
        return "ai"
    return message.type


def _role_to_message(role: str, content: str) -> BaseMessage:
    if role == "human":
        return HumanMessage(content=content)
    if role == "ai":
        return AIMessage(content=content)
    raise ValueError(f"unsupported message role: {role}")


class MinigrafChatMessageHistory(BaseChatMessageHistory):
    def __init__(self, session_id: str, path: str | Path | None = None) -> None:
        self.session_id = session_id
        self._session_slug = _session_slug(session_id)
        self._db = MiniGrafDb.open(str(path)) if path else MiniGrafDb.open_in_memory()

    @property
    def messages(self) -> list[BaseMessage]:
        session = _datalog_string(self.session_id)
        result = json.loads(
            self._db.execute(
                f"""(query [:find ?m ?index ?role ?content
                            :where [?m :message/session {session}]
                                   [?m :message/index ?index]
                                   [?m :message/role ?role]
                                   [?m :message/content ?content]])"""
            )
        )
        rows = sorted(result["results"], key=lambda row: row[1])
        return [_role_to_message(role, content) for _, _, role, content in rows]

    def add_messages(self, messages: Sequence[BaseMessage]) -> None:
        start = len(self.messages)
        facts = []
        for offset, message in enumerate(messages):
            index = start + offset
            entity = f":message/{self._session_slug}-{index}"
            role = _datalog_string(_message_to_role(message))
            content = _datalog_string(str(message.content))
            session = _datalog_string(self.session_id)
            facts.extend(
                [
                    f"[{entity} :message/session {session}]",
                    f"[{entity} :message/index {index}]",
                    f"[{entity} :message/role {role}]",
                    f"[{entity} :message/content {content}]",
                ]
            )
        if facts:
            self._db.execute(f"(transact [{' '.join(facts)}])")

    def clear(self) -> None:
        session = _datalog_string(self.session_id)
        result = json.loads(
            self._db.execute(
                f"""(query [:find ?m ?index ?role ?content
                            :where [?m :message/session {session}]
                                   [?m :message/index ?index]
                                   [?m :message/role ?role]
                                   [?m :message/content ?content]])"""
            )
        )
        facts = []
        for message_id, index, role, content in result["results"]:
            entity = f'#uuid "{message_id}"'
            facts.extend(
                [
                    f"[{entity} :message/session {session}]",
                    f"[{entity} :message/index {index}]",
                    f"[{entity} :message/role {_datalog_string(role)}]",
                    f"[{entity} :message/content {_datalog_string(content)}]",
                ]
            )
        if facts:
            self._db.execute(f"(retract [{' '.join(facts)}])")


def main() -> None:
    with tempfile.TemporaryDirectory() as directory:
        path = Path(directory) / "langchain-memory.mg"
        history = MinigrafChatMessageHistory("agent-demo", path)
        history.add_messages(
            [
                HumanMessage(content="Remember that Minigraf stores agent memory."),
                AIMessage(content="Got it. I will use Minigraf-backed chat history."),
            ]
        )

        for message in history.messages:
            label = "Human" if isinstance(message, HumanMessage) else "AI"
            print(f"{label}: {message.content}")


if __name__ == "__main__":
    main()
